### ISSUE TITLE
 Expected unicode, got quoted_name

### ISSUE THREAD

**Issue by TrentGlover** [dask/fastparquet#810]
Error returned when calling parquetobj.table_to_parquet(parq_file_name, row['schema_name'], row['table_name'] , sql, db_engine, batch_size)

error is 

Traceback (most recent call last):
  File "C:\Users\a093789\AppData\Roaming\Python\Python310\site-packages\fastparquet\writer.py", line 1499, in write_thrift
    return f.write(obj.to_bytes())
TypeError: Expected unicode, got quoted_name
TypeError: Expected unicode, got quoted_name
Exception ignored in: 'fastparquet.cencoding.write_list'

Code:

from tempfile import NamedTemporaryFile
import shutil, toml, csv, datetime, time, os, pysftp, pyodbc, fastparquet, sys
from datetime import timedelta, date
from dateutil.relativedelta import relativedelta
from dateutil import parser
import numpy as np
import pandas as pd
import paramiko
from paramiko.py3compat import decodebytes
from pysftp import Connection
from sqlalchemy import Table, create_engine, schema
import util
from os.path import exists
import time
import smtplib
from email.mime.text import MIMEText
import socket

class parquet():

   def __init__(self, conf):
       global confg
       confg = conf
       self.inststrttm = time.strftime("%Y%m%d_%H%M%S")
       self.btch_id = int(str(time.time()).split('.')[0])
       self.absmodpath = os.path.dirname(__file__)
       self.logrconf = os.path.join(self.absmodpath, 'loggers.conf')
       self.logger = util.getlogger(confg, self.logrconf, 'SNDEVNT', self.inststrttm, True)

   def __get_dtype(self, column, sqltype):
       import sqlalchemy.dialects as sqld
       from sqlalchemy.types import (TIMESTAMP, Boolean, Date, DateTime,
                                     Float, Integer)

       return str

   def __write_parquet(self, output_path: str, batch_array, column_dict, write_index: bool,
                       compression: str, append: bool):
       # Create the DataFrame to hold the batch array contents
       b_df = pd.DataFrame(batch_array, columns=column_dict, dtype=str)

       # Cast the DataFrame columns to the sqlalchemy column analogues
       b_df = b_df.astype(dtype=column_dict)

       # Write to the parquet file (first write needs append=False)
       if not b_df.empty:
           fastparquet.write(output_path, b_df,
                         write_index=write_index, compression=compression, append=append)

   def table_to_parquet(self, file_path: str, schema_name: str,
                        table_name: str, sql: str, db_engine,
                        batch_size: int = 10000, write_index: bool = False,
                        compression: str = 'GZIP'):

       self.logger.info("Process Started at {}".format(time.ctime()))
       self.logger.info("Process Batch ID - {}".format(self.btch_id))

       db_metadata = schema.MetaData(bind=db_engine, schema=schema_name)

       try:
           db_table = Table(table_name, db_metadata, autoload=True)

           # columns for the parquet file
           column_dict = dict()
           for column in db_table.columns:
               dtype = self.__get_dtype(column, column.type)
               column_dict[column.name] = dtype

           result = db_engine.execute(sql)
           self.logger.info("Executed SQL {}".format(sql))

           row_batch = result.fetchmany(size=batch_size)
           iteration = 1
           append = False

           self.logger.info("Initiating to write {} with batchsize {}".format(file_path, batch_size))

           while (len(row_batch) > 0):
               self.__write_parquet(file_path, row_batch,
                                    column_dict, write_index, compression, append)
               append = True
               self.logger.info("Iteration number {}".format(iteration))
               row_batch = result.fetchmany(size=batch_size)
               iteration = iteration + 1

       except ValueError as e:
           self.logger.critical("Error occured while trying to create parquet files {}".format(e.__str__))

       self.logger.info("Parquet file writing process completed successfully.")
#

if __name__ == '__main__':

    sender = 'bi-pssis@mwiah.com'
    receivers = ['tglover@mwiah.com', '2082833572@tmomail.net']
    port = 25

    conf = toml.load('config/params.toml')

    constr = "DRIVER={SQL Server Native Client 11.0};SERVER=bi-prtr;DATABASE=archive;Trusted_Connection=yes;"

    parqfilepath = 'C:\\appl\Python\Data\Parquet\\'
    parqfilename = 'raw-cna-mwia-archive-archive_{}-interval_{}_{}_dumpedon_{}.parquet'
    parqerrorfilepath = '\\\\cake\\DWBackup\EAP-Archive\\data\\parquet\\'
    batch_size = 1000000

    tempfile = NamedTemporaryFile(mode='w', delete=False, newline='', dir='./')
    number_of_rows = 0

    conf = toml.load('config/params.toml')
    parquetobj = parquet(conf)

    continue_processing: bool = True
    todays_date = date.today()

    line_count = 1
    from_date = date.today()
    to_date = date.today()

    try:
        conn = pyodbc.connect(constr)
    except:
        print("{c} is not working".format(c=constr))

    cursor = conn.cursor()
    db_engine = create_engine('mssql+pyodbc:///?odbc_connect=' + constr)
    sql = 'select * from archive.Ibs.PHOALD WITH (NOLOCK) Where Convert(date, Transaction_time) > Convert(date, \'2019-05-20\') and Convert(date, Transaction_time) <= Convert(date, \'2019-06-20\')'
    print(f"{sql}")

    parq_file_name = 'testfile.parquet'

    parquetobj.table_to_parquet(parq_file_name, 'IBS', 'PHOALD', sql, db_engine,
                                                batch_size)

- Dask version: 2022.9.1
- Python version: 3.10
- Operating System: Windows 10 Enterprise version 10.0.19041 Build 19041
- Install method (conda, pip, source): pip

**Comment by martindurant** [comment]
That's a lot of code and stuff I'm not familiar with, it would be really useful to try to pare down the reproducer to the minimum required to show the problem.

At a complete guess `columns=column_dict` in `pd.DataFrame()` is passing some non-string items derived from the database stuff. You may need explicit `str()` around these.

**Comment by TrentGlover** [comment]
Yeah, I’m really trying to trim down the code to the minimum required. This is my first rodeo. I’ll get back to you.

Regards,

Trent Glover
AmerisourceBergen
MWI Animal Health
Data Engineer

Work Phone: 208.955.9527
Cell/text: 208.283.3572

3041 W Pasadena Dr.
Boise, ID 83705
United States
www.mwiah.com

United in our responsibility
to create healthier futures

Confidentiality notice: This electronic mail transmission may contain privileged, confidential and/or protected personal information and is intended only for the review of the party to whom it is addressed. Any unauthorized use or disclosure of the information contained herein may be a violation of applicable law. If you have received this transmission in error, please immediately return it to the sender, delete it and destroy it without reading it. Unintended transmission shall not constitute the waiver of the attorney-client or any other privilege.

**Comment by TrentGlover** [comment]
same error when attempting to turn all columns into strings...

data_query: str = "Select {} from archive.Ibs.PHOALD WITH (NOLOCK) Where Convert(date, Transaction_time) > Convert(" \
                      "date, \'2019-05-20\') and Convert(date, Transaction_time) <= Convert(date, \'2019-06-20\') "

...

    cursor = conn.cursor()
    metadata_sql = 'select top 1 * from archive.Ibs.PHOALD WITH (NOLOCK)'
    print(metadata_sql)
    cursor.execute(metadata_sql)
    column_names = ''
    field_names = [i[0] for i in cursor.description]
    for j in field_names:
        j = 'convert(nvarchar, ' + j + ') as ' + j
        column_names = column_names + ',' + j
    column_names = column_names[1:]
    print(f"{column_names}")

    sql = data_query.format(column_names)
    print(f"{sql}")

**Comment by martindurant** [comment]
A "quoted_name" is entering into your dataframe metadata *somewhere*, so the task is to find out where. Maybe fastparquet should be converting it to a string internally, but I would bet that requiring a real string is a reasonable expectation in this case.

I'm afraid I don't really know what your code it doing, I don't normally work with DB cursor objects. You might find it easier to use pd.from_sql_query instead, as I suspect it might do the correct conversions for you.

**Comment by TrentGlover** [comment]
the problem definitely looks like it is native to the native code

**Comment by martindurant** [comment]
> i've called it without modification and it is still throwing the error.

What did you call exactly? Did you manage to make a dataframe showing this problem without SQL?

**Comment by TrentGlover** [comment]
The construction and the population of the dataframes are within the __write_parquet function which is called from the table_to_parquet function within the fastparquet code. I have not attempted to build a dataframe outside of this code.

I’m new to all of this so I apologize if I don’t get exactly how this works, but I’m assuming that unless I have something useful to add, I really shouldn’t be messing with the code that is part of the offering.

Regards,

Trent Glover
AmerisourceBergen
MWI Animal Health
Data Engineer

Work Phone: 208.955.9527
Cell/text: 208.283.3572

3041 W Pasadena Dr.
Boise, ID 83705
United States
www.mwiah.com

United in our responsibility
to create healthier futures

Confidentiality notice: This electronic mail transmission may contain privileged, confidential and/or protected personal information and is intended only for the review of the party to whom it is addressed. Any unauthorized use or disclosure of the information contained herein may be a violation of applicable law. If you have received this transmission in error, please immediately return it to the sender, delete it and destroy it without reading it. Unintended transmission shall not constitute the waiver of the attorney-client or any other privilege.

**Comment by martindurant** [comment]
I mean:
```
     b_df = pd.DataFrame(row_batch, columns=column_dict, dtype=str)
     b_df = b_df.astype(dtype=column_dict)
```
here both the original `columns` and the ones supplied in `dtype=` have keys that came from the SQL driver, and probably responsible for the "quoted_name" in the error. Try converting them to actual str before doing anything with them:
```
columns = {str(k): v for k, v in columns_dict.items()}
```

Are the dtypes really all str?
